Mathlib/Probability/Hoeffding.lean

```diff
@@ -0,0 +1,233 @@
+    integrableExpSet X μ = Set.univ := by
+  apply Set.eq_univ_iff_forall.mpr
+  intro t
+  by_cases sign_t : 0 ≤ t
+  case pos =>
+    apply integrable_exp_of_ae_le_const _ _ hX
+      (by filter_upwards [hb] with ω hb using hb.2) _ sign_t
+  case neg =>
+    apply integrable_exp_of_ae_const_le _ _  hX
+      (by filter_upwards [hb] with ω hb using hb.1) _ (le_of_not_ge sign_t)
+
+theorem integrable_exp_set_interior_of_ae_mem_Icc  [IsFiniteMeasure μ] {X : Ω → ℝ} (t a b : ℝ)
+    (hX : AEMeasurable X μ)
+    (h : ∀ᵐ (ω : Ω) ∂μ, X ω ∈ Set.Icc a b) : t ∈ interior (integrableExpSet X μ) := by
+  rw [integrable_exp_of_ae_mem_Icc μ a b hX h]
+  apply mem_interior_iff_mem_nhds.mpr Filter.univ_mem
+
+theorem cgf_deriv_eq_tilted_measure_expectation {X : Ω → ℝ} (t : ℝ)
+    (h0 : t ∈ interior (integrableExpSet X μ)) :
+    deriv (cgf X μ) t = (μ.tilted fun ω ↦ t * X ω)[X] := by
+  rw [deriv_cgf]
+  · rw [MeasureTheory.integral_tilted, ← integral_div]
+    simp only [smul_eq_mul, mgf]; congr with ω; ring
+  · exact h0
+
+theorem cgf_iterated_deriv_two_eq_tilted_measure_variance [IsProbabilityMeasure μ]
+    {X : Ω → ℝ} (t a b : ℝ) (hX : AEMeasurable X μ) (h : ∀ᵐ (ω : Ω) ∂μ, X ω ∈ Set.Icc a b):
+    iteratedDeriv 2 (cgf X μ) t = Var[X; μ.tilted fun ω ↦ t * X ω] := by
+  rw [iteratedDeriv_two_cgf]
+  · have p : (μ.tilted (fun ω ↦ t * X ω))[fun ω ↦ (X ω) ^ 2] =
+      (μ[fun ω ↦ (X ω) ^ 2 * rexp (t * X ω)]) / mgf X μ t := by
+      rw [MeasureTheory.integral_tilted, ← integral_div]
+      simp only [smul_eq_mul, mgf]; congr with ω; ring
+    rw [<- p, cgf_deriv_eq_tilted_measure_expectation _ t
+      (integrable_exp_set_interior_of_ae_mem_Icc μ t a b hX h)]
+    dsimp only
+    have p : Var[X; μ.tilted fun ω ↦ t * X ω] =
+        (μ.tilted fun ω ↦ t * X ω)[X ^ 2] - ((μ.tilted fun ω ↦ t * X ω)[X]) ^ 2 := by
+      have _ : IsProbabilityMeasure (μ.tilted fun ω ↦ t * X ω) :=
+        isProbabilityMeasure_tilted
+        (by rw [integrable_exp_of_ae_mem_Icc μ a b hX h]; exact trivial : t ∈ integrableExpSet X μ)
+      have hμ := tilted_absolutelyContinuous μ fun ω ↦ t * X ω
+      apply variance_def' <|
+        memLp_of_bounded (hμ h) (AEMeasurable.aestronglyMeasurable (hX.mono_ac hμ)) 2
+    rw [p]; exact rfl
+  · exact integrable_exp_set_interior_of_ae_mem_Icc μ t a b hX h
+
+theorem cgf_le_bound_of_ae_mem_Icc_and_mean_zero [IsProbabilityMeasure μ]
+    (t a b : ℝ) {X : Ω → ℝ} (ht : 0 < t)
+    (hX : AEMeasurable X μ)
+    (h : ∀ᵐ (ω : Ω) ∂μ, X ω ∈ Set.Icc a b) (h0 : ∫ (x : Ω), X x ∂μ = 0):
+  cgf X μ t ≤ t ^ 2 * (b - a) ^ 2 / 8 := by
+  let f := cgf X μ
+  let f' := deriv (cgf X μ)
+  let f'' := iteratedDeriv 2 (cgf X μ)
+  have hf : f 0 = 0 := cgf_zero
+  have hf' : f' 0 = 0 := by
+    dsimp [f']
+    rw [deriv_cgf_zero]
+    · simp only [measure_univ, ENNReal.one_toReal, div_one, f']
+      exact h0
+    · exact integrable_exp_set_interior_of_ae_mem_Icc μ 0 a b hX h
+  have q : ∃ c ∈ (Set.Ioo 0 t), f t = f 0 + f' 0 * t + f'' c * t ^ 2 / 2 := by
```

Comment:
As I wrote in a previous comment, I still think that you should extract a lemma here. Please write a lemma about a generic real function f with derivatives f' and f''. This is a very useful particular case of Taylor's theorem and merits it's own lemma, especially since it's apparently not so easily derived from the more general versions in Mathlib.